crates/biome_js_analyze/src/lint/nursery/no_shadow.rs

```diff
@@ -0,0 +1,227 @@
+            // uncomment when we can handle the test cases from typescript-eslint
+            // RuleSource::EslintTypeScript("no-shadow"),
+        ],
+    }
+}
+
+pub struct ShadowedBinding {
+    /// The binding that is violating the rule.
+    binding: Binding,
+    /// The binding that is shadowed.
+    shadowed_binding: Binding,
+}
+
+impl Rule for NoShadow {
+    type Query = SemanticServices;
+    type State = ShadowedBinding;
+    type Signals = Box<[Self::State]>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let mut shadowed_bindings = Vec::new();
+        let model = ctx.query();
+
+        for binding in ctx.query().all_bindings() {
+            if let Some(shadowed_binding) = check_shadowing(model, binding) {
+                shadowed_bindings.push(shadowed_binding);
+            }
+        }
+
+        shadowed_bindings.into_boxed_slice()
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        //
+        // Read our guidelines to write great diagnostics:
+        // https://docs.rs/biome_analyze/latest/biome_analyze/#what-a-rule-should-say-to-the-user
+        //
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                state.binding.tree().range(),
+                markup! {
+                    "This variable shadows another variable with the same name in the outer scope."
+                },
+            )
+            .detail(
+                state.shadowed_binding.tree().range(),
+                markup!(
+                    "This is the shadowed variable, which is now inaccessible in the inner scope."
+                ),
+            )
+            .note(markup! {
+                "Consider renaming this variable. It's easy to confuse the origin of variables if they share the same name."
+            }),
+        )
+    }
+}
+
+fn check_shadowing(model: &SemanticModel, binding: Binding) -> Option<ShadowedBinding> {
+    let name = get_binding_name(&binding)?;
+
+    for upper in binding.scope().ancestors().skip(1) {
+        if let Some(upper_binding) = upper.get_binding(name.clone()) {
+            if binding.syntax() == upper_binding.syntax() {
+                // a binding can't shadow itself
+                continue;
+            }
+            if is_on_initializer(&binding, &upper_binding) {
+                continue;
+            }
+            if is_redeclaration(model, &binding, &upper_binding) {
+                // redeclarations are not shadowing, they get caught by `noRedeclare`
+                continue;
+            }
+            if is_declaration(&binding)
+                && is_declaration(&upper_binding)
+                && upper_binding.syntax().text_range().start() >= binding.scope().range().end()
+            {
+                // the shadowed binding must be declared before the shadowing one
+                continue;
+            }
```

Comment:
This check is probably the most dubious one here because I don't have a source for it. I'm not entirely sure which part of the JS spec would be relevant.